datastruct/mpool.h

```diff
@@ -11,69 +13,114 @@
  * of allowing the memory to be reused by some other code.
  */
 
+struct mpool {
+	size_t stacklen;
+	size_t allocsize;
+	void ** allocs;
+	uint64_t nallocs;
+	uint64_t nempties;
+	int state;
+	void ** allocs_static;
+	void (*atexitfunc)(void);
+};
+
+static inline void
+mpool_atexit(struct mpool * M)
+{
+
+	while (M->stacklen)
+		free(M->allocs[--M->stacklen]);
+	if (M->allocs != M->allocs_static)
+		free(M->allocs);
+}
+
+static inline void *
+mpool_malloc(struct mpool * M, size_t len)
+{
+
+	M->nallocs++;
+	if (M->stacklen)
+		return (M->allocs[--(M->stacklen)]);
+	M->nempties++;
+	if (M->state == 0) {
+		atexit(M->atexitfunc);
+		M->state = 1;
+	}
+	return (malloc(len));
+}
+
+static inline void
+mpool_free(struct mpool * M, void * p)
+{
+	void * allocs_new;
+
+	if (p == NULL)
+		return;
+
+	if (M->stacklen < M->allocsize) {
+		M->allocs[M->stacklen++] = p;
+		return;
+	}
+
+	if (M->nempties	> (M->nallocs >> 8)) {
+		allocs_new = malloc(M->allocsize * 2 * sizeof(void *));
+		if (allocs_new) {
+			memcpy(allocs_new, M->allocs,
+			    M->allocsize * 2 * sizeof(void *));
+			if (M->allocs != M->allocs_static)
+				free(M->allocs);
+			M->allocs = allocs_new;
+			M->allocsize = M->allocsize * 2;
+			M->allocs[M->stacklen++] = p;
+		} else
+			free(p);
+	} else
+		free(p);
+	M->nempties = 0;
+	M->nallocs = 0;
+}
+
 /**
  * MPOOL(name, type, size):
  * Define the functions
  *
  * ${type} * mpool_${name}_malloc(void);
  * void mpool_${name}_free(${type} *);
  *
- * which allocate and free structures of type ${type}.  Up to ${size}
- * such structures are kept cached after _free is called in order to
- * allow future _malloc calls to be rapidly serviced.
+ * which allocate and free structures of type ${type}.  A minimum of ${size}
+ * such structures are kept cached after _free is called in order to allow
+ * future _malloc calls to be rapidly serviced; this limit will be autotuned
+ * upwards depending on the allocation/free pattern.
  *
  * Cached structures will be freed at program exit time in order to aid
  * in the detection of memory leaks.
  */
-#define MPOOL(name, type, size) 				\
-static struct mpool_##name##_struct {				\
-	size_t stacklen;					\
-	int atexit_set;						\
-	type * allocs[size];					\
-} mpool_##name##_rec = {0, 0};					\
+#define MPOOL(name, type, size)					\
+static void mpool_##name##_atexit(void);			\
+static void * mpool_##name##_static[size];			\
+static struct mpool mpool_##name##_rec =			\
+    {0, size, mpool_##name##_static, 0, 0, 0,			\
```

Comment:
This sets `M->stacklen = 0`.  Shouldn't this be `size`?  Because we created `static void * mpool_##name##_static[size];` ?

Reply:
Never mind, after going back to the commit that first introduces the stack, I now see that it shouldn't be non-zero.

Reply:
Right, `stacklen` says how many pointers-to-allocations we have.  We start out with zero such allocations, but space to store some if they arrive via mpool_free.